File: src/main/java/com/sedmelluq/discord/lavaplayer/track/TrackStateListener.java
package com.sedmelluq.discord.lavaplayer.track;

import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;

/**
 * Listener of track execution events.
 */
public interface TrackStateListener {
  /**
   * Called when an exception occurs while a track is playing or loading. This is always fatal, but it may have left
   * some data in the audio buffer which can still play until the buffer clears out.
   *
   * @param track The audio track for which the exception occurred
   * @param exception The exception that occurred
   */
  void onTrackException(AudioTrack track, FriendlyException exception);

  /**
   * Called when an exception occurs while a track is playing or loading. This is always fatal, but it may have left
   * some data in the audio buffer which can still play until the buffer clears out.
   *
   * @param track The audio track for which the exception occurred
   * @param thresholdMs The wait threshold that was exceeded for this event to trigger
   */
  void onTrackStuck(AudioTrack track, long thresholdMs);
}

File: main/src/main/java/lavaplayer/track/TrackStateListener.kt
package lavaplayer.track

import lavaplayer.tools.FriendlyException

/**
 * Listener of track execution events.
 */
interface TrackStateListener {
    /**
     * Called when an exception occurs while a track is playing or loading. This is always fatal, but it may have left
     * some data in the audio buffer which can still play until the buffer clears out.
     *
     * @param track     The audio track for which the exception occurred
     * @param exception The exception that occurred
     */
    fun onTrackException(track: AudioTrack, exception: FriendlyException)

    /**
     * Called when an exception occurs while a track is playing or loading. This is always fatal, but it may have left
     * some data in the audio buffer which can still play until the buffer clears out.
     *
     * @param track       The audio track for which the exception occurred
     * @param thresholdMs The wait threshold that was exceeded for this event to trigger
     */
    fun onTrackStuck(track: AudioTrack, thresholdMs: Long)
}
